### PR TITLE
Add ability to manage rules with the ServiceBusReceiver

### PR DESCRIPTION
**Pull request by ZejiaJiang**
# Description

Add new async client responsible for managing rules for a specific topic subscription. The rule manager
 requires only Listen claims, whereas the ServiceBusAdministrationAsyncClient requires Manage claims.
Addresses https://github.com/Azure/azure-sdk-for-java/issues/27711

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by conniey** [comment]
@srnagar , can you take a look at the API view for this? .NET, JS have shipped their RuleManagers.. so we're trying to ship the same experience.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:482]: Iterator might be a better pattern in this code (than count-- and still index by hardcoded number).

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:505]: same here

is this code copied from somewhere?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:137]: No, the CreateRuleOptions constructure check it. forget add this exception in comment. will add it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:148]: Just check the admin client and .net client, it return a list, Mono<List> may be better, will change it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:515]: The .net SDK didn't support input options. Just keep same parameter with .net SDK

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:136]: will fix it!

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:652]: will fix it!

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:505]: Yeah, these two util functions are copied from Track 1 SDK, I'll do a better pattern and remove hardcode.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:116]: forgot add filter is null caused exception comment, action could be null.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:447]: Yeah, will add a check for it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (weidongxu-microsoft) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:116]: there are something like this (besides null check) https://github.com/Azure/azure-sdk-for-java/blob/048b067290bfeab162ebdf3188998c62a8878ddb/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java#L411 ?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (weidongxu-microsoft) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:505]: You can wait for more inputs from Connie or Srikanta before any change, as the coding style is highly opinionated, and they might not agree with me.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:672]: cast exception and if when decode a unexpecpted filter will throw UnsupportedOperationException, I wrapped them as a amqp exception.

refer to RejectedExecutionException throwed in the ReactorDispatcher class, it just add the error message to format string, I'll fix like this first.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusRuleManagerAsyncClient.java:104]: nit: For the sake of consistency, also checking Objects.requireNotNull

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/administration/models/RuleProperties.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/administration/models/RuleProperties.java:214]: We didn't want to add a public constructor for this so that's why we have the EntityHelper with the EntityRuleAccessor: https://github.com/Azure/azure-sdk-for-java/blob/3ec8fccab55cfaf7428ed9424f57df1bdb4132d6/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/administration/models/RuleProperties.java#L38

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (conniey) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:671]: Could you copy that response to: https://github.com/Azure/azure-sdk-for-java/pull/30271/files#diff-cb1df7ed3367cd376e665cee3e4105c159ca2d6f85c0471d2ffafbea88503a6fR650 ? I think that would help set some context for future eyes.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:438]: We should use AmqpResponseCode here to check messageStatus

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (conniey) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:482]: I agree with weidong, this is a bit hard to follow.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (conniey) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:505]: I agree with weidong on this. It is a bit hard to follow.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (anuchandy) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:478]: Please check other similar places in this pr where we use HashMap

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:509]: The reactive way would be to return `Flux<RuleProperties>`.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (anuchandy) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:674]: I wonder if we really have to wrap this terminal exception as a retriable AmqpException?, or just throwing as it is is good enough \\cc @conniey 

(The reason for the mapping exception in ReactorDispatcher to AmqpException was the library to retry internally)

**Comment by joshfree** [comment]
@ki1729 can you please review this SB pr?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java
Comment (conniey) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ManagementChannel.java:674]: I don't think we have to wrap this exception.

**Review by anuchandy — APPROVED** [review]
Thanks, Zeija!, Looks good to me but see if other reviewers have any more feedback. Also, we need to get API signoff from Srikanta

**Review by ki1729 — APPROVED** [review]

**Comment by ZejiaJiang** [comment]
Hey @JonathanGiles ,could you please take a look at the API view for this? .NET, JS have shipped their RuleManagers.. so we're trying to ship the same experience.

**Comment by JonathanGiles** [comment]
> Hey @JonathanGiles ,could you please take a look at the API view for this? .NET, JS have shipped their RuleManagers.. so we're trying to ship the same experience.

I've left one comment for your consideration. Could you also please share a link to the .NET equivalent API, so that I may compare? Thanks.

**Comment by ZejiaJiang** [comment]
@JonathanGiles 
Thank you, and here is .net related information:
.NET issue - https://github.com/Azure/azure-sdk-for-net/issues/25523
.NET PR - https://github.com/Azure/azure-sdk-for-net/pull/27346

**Comment by JonathanGiles** [comment]
Why is the API in Java `createRule` and `deleteRule` and in .NET it is `addRule` and `removeRule`? Is there precedence that we are following here? It seems that what we have in Java is consistent internally at least.

**Comment by ZejiaJiang** [comment]
> Why is the API in Java `createRule` and `deleteRule` and in .NET it is `addRule` and `removeRule`? Is there precedence that we are following here? It seems that what we have in Java is consistent internally at least.

Sorry, I forgot to add the latest PR https://github.com/Azure/azure-sdk-for-net/pull/29407 , in this PR , .net change `addRule` and `removeRule` to `createRule` and `deleteRule`.

**Comment by ZejiaJiang** [comment]
Hi @JonathanGiles , I added another PR above to explain why using the `createRule` and `deleteRule`, here is the source code of the rule manager : https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/RuleManager/ServiceBusRuleManager.cs you can compare the final version, in case that .NET did some change after the first PR.

**Comment by JonathanGiles** [comment]
This looks good to me. I left one API comment, and I would like to make sure the Service Bus Java folks are all happy with this too before I hit approve. Thanks!

**Comment by ZejiaJiang** [comment]
Jonathon suggest use List in api.dev, refer to other api which return flux in async client, they return IterableStream in sync client, so   I return IterableStream instead of List.

@ki1729 @conniey @anuchandy Please take a look again, if all is OK, I can ask @JonathanGiles approve this PR.

**Review by anuchandy — APPROVED** [review]
@ZejiaJiang usage of `IterableSteam` for list is the right sync api pattern.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:523]: Is returning null a normal condition? (if so, just making sure calling code has a null check

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:543]: Same as comment above, just making sure we don't accidentally run into a NPE

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java:535]: I don't think these tests are enough. It shows that the method doesn't throw... we may want to assert things like.. did we call the REquestResponse channel with properties I expected?

Here's an example:
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java#L163-L184

Similarly with the ones below...

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java:540]: It would be worthwhile to create a message _with_ some rules so you can check it deserialized properly.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ManagementChannelTests.java:548]: Same with comment above about checking if we made the right delete call.

**Review by conniey — APPROVED** [review]
Thanks so much! I just have a few comments about the tests. 🚢

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:543]: Checked, and we allow a null rule action.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java
Comment (ZejiaJiang) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:523]: Already has a null check in calling code.

return null won't reach if service response correctly.

**Comment by ZejiaJiang** [comment]
Hi @JonathanGiles , I improve the tests following Connie's comments. Could you please take a look again? Thank you!

**Pull request merged**
